java/runtime/src/main/java/io/ray/runtime/metric/Metric.java

```diff
@@ -0,0 +1,98 @@
+package io.ray.runtime.metric;
+
+import com.google.common.base.Preconditions;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * Class metric is mapped to stats metric object in core worker.
+ * it must be in categories set [Gague, Count, Sum, Histogram].
+ */
+public abstract class Metric {
+  protected String name;
+
+  protected double value;
+  // Native pointer mapping to gauge object of stats.
+  protected long metricNativePointer = 0L;
+
+  protected Map<TagKey, String> tags;
+
+  public Metric(String name, Map<TagKey, String> tags) {
+    Preconditions.checkNotNull(tags, "Metric tags map must not be null.");
+    Preconditions.checkNotNull(name, "Metric name must not be null.");
+    this.name = name;
+    this.tags = tags;
+    this.value = 0.0d;
+  }
+
+  // Sync metric with core worker stats for registry.
+  // Metric data will be flushed into stats view data inside core worker immediately after
+  // record is called.
+  /**
+   * Flush records to stats in last aggregator.
+   */
+  public void record() {
+    Preconditions.checkState(metricNativePointer != 0, "Metric native pointer must not be 0.");
+    // Get tag key list from map;
+    List<TagKey> nativeTagKeyList = new ArrayList<>();
+    List<String> tagValues = new ArrayList<>();
+    for (Map.Entry<TagKey, String> entry : tags.entrySet()) {
+      nativeTagKeyList.add(entry.getKey());
+      tagValues.add(entry.getValue());
+    }
+    // Get tag value list from map;
+    recordNative(metricNativePointer, value, nativeTagKeyList.stream()
+        .map(TagKey::getTagKey).collect(Collectors.toList()), tagValues);
+  }
+
+  private native void recordNative(long gaugePtr, double value,
```

Comment:
Maybe we should use `metricPtr` instead of `gaugePtr`

Reply:
> Maybe we should use `metricPtr` instead of `gaugePtr`

typo, I'll fix later.